mlxbf-bootctl.c

```diff
@@ -310,22 +342,73 @@ FILE *open_sysfs(const char *name, const char *attr)
   return f;
 }
 
-int get_watchdog(void)
+uint16_t get_watchdog_interval(void)
 {
+  uint16_t interval;
+  uint32_t watchdog;
   FILE *f = open_sysfs(POST_RESET_WDOG_PATH, "r");
-  int watchdog;
+
   if (fscanf(f, "%d", &watchdog) != 1)
     die("%s: failed to read integer", POST_RESET_WDOG_PATH);
   fclose(f);
-  return watchdog;
+
+  interval = (watchdog >> POST_RESET_WDOG_INTERVAL_SHIFT) & POST_RESET_WDOG_INTERVAL_MASK;
+  return interval;
 }
 
-void set_watchdog(int interval)
+uint8_t get_watchdog_mode(void)
 {
-  FILE *f = open_sysfs(POST_RESET_WDOG_PATH, "w");
-  if (fprintf(f, "%d\n", interval) < 0)
-    die("%s: failed to set watchdog to '%d'", POST_RESET_WDOG_PATH, interval);
+  uint16_t mode;
+  uint32_t watchdog;
+  FILE *f = open_sysfs(POST_RESET_WDOG_PATH, "r");
+
+  if (fscanf(f, "%d", &watchdog) != 1)
+    die("%s: failed to read integer", POST_RESET_WDOG_PATH);
+  fclose(f);
+
+  mode = (watchdog >> POST_RESET_WDOG_MODE_SHIFT) & POST_RESET_WDOG_MODE_MASK;
+  return mode;
+}
+
+const char *get_watchdog_mode_str(void)
+{
+  uint16_t mode = get_watchdog_mode();
+  switch (mode) {
+    case BF_BOOT_WDOG_MODE_DISABLED:
+      return BF_BOOT_WDOG_MODE_DISABLED_STR;
+    case BF_BOOT_WDOG_MODE_STANDARD:
+      return BF_BOOT_WDOG_MODE_STANDARD_STR;
+    case BF_BOOT_WDOG_MODE_TIME_LIMIT:
+      return BF_BOOT_WDOG_MODE_TIME_LIMIT_STR;
+    default:
+      return BF_BOOT_WDOG_MODE_UNKNOWN_STR;
+  }
+}
+
+void set_watchdog(uint16_t interval, uint8_t mode)
+{
+  FILE *f = NULL;
+  uint32_t watchdog = 0;
+
+  if (mode > BF_BOOT_WDOG_MODE_MAX) {
+    die("Watchdog mode '%d' is unsupported.", mode);
+  }
+
+  if (interval != 0 && interval < MIN_WDOG_INTERVAL_SEC) {
+    die("Watchdog interval '%d' is too small (must be >= %ds).", interval, MIN_WDOG_INTERVAL_SEC);
+  }
+
+  watchdog |= (interval & POST_RESET_WDOG_INTERVAL_MASK) << POST_RESET_WDOG_INTERVAL_SHIFT;
+  watchdog |= (mode & POST_RESET_WDOG_MODE_MASK) << POST_RESET_WDOG_MODE_SHIFT;
+
+  f = open_sysfs(POST_RESET_WDOG_PATH, "w");
+  if (fprintf(f, "%d\n", watchdog) < 0)
+    die("%s: failed to set watchdog to '0x%X'", POST_RESET_WDOG_PATH, watchdog);
   fclose(f);
+
+  if (mode > 0 && interval != get_watchdog_interval()) {
+    die("Failed to set watchdog due to outdated ATF version.");
```

Comment:
For outdated ATF, should we compare mode instead (since interval is compatible)?

Reply:
Yeah, good point. The above works in most cases except for the case where boot swap watchdog was set with an interval and then we try to set a new mode with the same interval (because mode will always return 0 in older ATF).

It's actually probably good to still check both the mode and interval to make sure the whole operation worked as expected - will update.